diff --git a/node_modules/@react-native-firebase/database/lib/DatabaseSyncTree.js b/node_modules/@react-native-firebase/database/lib/DatabaseSyncTree.js
index 59d4653..5088ccc 100644
--- a/node_modules/@react-native-firebase/database/lib/DatabaseSyncTree.js
+++ b/node_modules/@react-native-firebase/database/lib/DatabaseSyncTree.js
@@ -164,15 +164,18 @@ class DatabaseSyncTree {
       let subscriptions;
 
       // EventEmitter in react-native < 0.70 had a `_subscriber` property with a method for subscriptions by type...
-      if (SharedEventEmitter._subscriber) {
-        subscriptions = SharedEventEmitter._subscriber.getSubscriptionsForType(registration);
-      } else {
-        // ...react-native 0.70 now stores subscribers as a map of Sets by type in `_registry`
-        const registrySubscriptionsSet = SharedEventEmitter._registry[registration];
-        if (registrySubscriptionsSet) {
-          subscriptions = Array.from(registrySubscriptionsSet);
+        if (SharedEventEmitter) {
+            if (SharedEventEmitter._subscriber) {
+                subscriptions = SharedEventEmitter._subscriber.getSubscriptionsForType(registration);
+            }
+            if (SharedEventEmitter._registry){
+                // ...react-native 0.70 now stores subscribers as a map of Sets by type in `_registry`
+                const registrySubscriptionsSet = SharedEventEmitter._registry[registration];
+                if (registrySubscriptionsSet) {
+                    subscriptions = Array.from(registrySubscriptionsSet);
+                }
+            }
         }
-      }
 
       if (subscriptions) {
         for (let j = 0, l = subscriptions.length; j < l; j++) {
