patch-package
--- a/node_modules/react-native-showdown/lib/MarkdownView.js
+++ b/node_modules/react-native-showdown/lib/MarkdownView.js
@@ -1,5 +1,6 @@
 
-import React, { Component, View, Text, WebView, StyleSheet, LinkingIOS } from 'react-native';
+import React, { Component,} from 'react';
+import { View, Text, WebView, StyleSheet, Linking } from 'react-native';
 import showdown from 'showdown';
 
 import defaultHTML from './defaultHTML';
@@ -25,26 +26,39 @@ class MarkdownView extends Component {
 		const { navigationStateChange } = this.state;
 
 		const html = defaultHTML
-				.replace('$title', '')
-				.replace('$body', this.converter.makeHtml(body))
-				.replace('$pureCSS', pureCSS);
+			.replace('$title', '')
+			.replace('$body', this.converter.makeHtml(body))
+			.replace('$pureCSS', pureCSS);
 
 		return <WebView
-				html={html + '<!-- ' + navigationStateChange + ' -->'}
-				automaticallyAdjustContentInsets={ automaticallyAdjustContentInsets }
-				onNavigationStateChange={ this.onNavigationStateChange.bind(this) }
-				style={ style } />;
+			source={{ html: html + '' }}
+			automaticallyAdjustContentInsets={ automaticallyAdjustContentInsets }
+			onNavigationStateChange={ this.onNavigationStateChange.bind(this) }
+			style={ style } />;
 	}
-
 	onNavigationStateChange(navState) {
 		// Check if user pressed on a link
-		if(navState.url !== 'about:blank' && navState.navigationType !== WebView.NavigationType.other) {
-			LinkingIOS.openURL(navState.url);
+		//check webview navigationType on android ?
+		if (navState.url !== 'about:blank' && !navState.url.startsWith("data:text/html")&& !navState.url.startsWith("http")) {
+			Linking.openURL(navState.url);
 			this.setState({
 				navigationStateChange: this.state.navigationStateChange + 1
 			});
 		}
+
+
 	}
+	/*
+	 onNavigationStateChange(navState) {
+	 // Check if user pressed on a link
+	 if (navState.url !== 'about:blank' && navState.navigationType !== WebView.NavigationType.other) {
+	 Linking.openURL(navState.url);
+	 this.setState({
+	 navigationStateChange: this.state.navigationStateChange + 1
+	 });
+	 }
+	 }
+	 */
 }
 
 MarkdownView.propTypes = {
--- a/node_modules/react-native-showdown/lib/defaultHTML.js
+++ b/node_modules/react-native-showdown/lib/defaultHTML.js
@@ -9,6 +9,7 @@ module.exports = `<!DOCTYPE html>
 		<style type="text/css">
 			body {
 				font-family: Roboto, '-apple-system', Helvetica Neue, Arial;
+				backgroundColor: '#FFFFFF';
 			}
 			b, strong {
 				font-family: Roboto, '-apple-system', Helvetica Neue, Arial;
