diff --git a/node_modules/react-native-linear-gradient/BVLinearGradient.podspec b/node_modules/react-native-linear-gradient/BVLinearGradient.podspec
index 5700e03..38f171d 100644
--- a/node_modules/react-native-linear-gradient/BVLinearGradient.podspec
+++ b/node_modules/react-native-linear-gradient/BVLinearGradient.podspec
@@ -16,6 +16,6 @@ Pod::Spec.new do |s|
   s.preserve_paths  = "**/*.js"
   s.frameworks = 'UIKit', 'QuartzCore', 'Foundation'
 
-  s.dependency 'React'
+  s.dependency 'React-Core'
 
 end
