patch-package
--- a/node_modules/react-native-progress-bar/ProgressBar.js
+++ b/node_modules/react-native-progress-bar/ProgressBar.js
@@ -1,21 +1,15 @@
-var React = require('react-native');
-
-var {
-  Animated,
-  Easing,
-  StyleSheet,
-  View
-} = React;
+import React, { Component,} from 'react';
+import { Animated, Easing, StyleSheet, View } from 'react-native';
 
 var styles = StyleSheet.create({
   background: {
     backgroundColor: '#bbbbbb',
-    height: 5,
+    height: 8,
     overflow: 'hidden'
   },
   fill: {
     backgroundColor: '#3b5998',
-    height: 5
+    height: 8
   }
 });
 
