patch-package
--- a/node_modules/react-native-markdown/Markdown.js
+++ b/node_modules/react-native-markdown/Markdown.js
@@ -1,7 +1,7 @@
-var React = require('react-native');
-var {
+import React from 'react';
+import {
   View
-} = React;
+} from 'react-native';
 var _ = require('lodash');
 var SimpleMarkdown = require('simple-markdown');
 
--- a/node_modules/react-native-markdown/rules.js
+++ b/node_modules/react-native-markdown/rules.js
@@ -1,9 +1,9 @@
-var React = require('react-native');
-var {
-  Image,
-  Text,
-  View,
-} = React;
+import React from 'react';
+import {
+    View,
+    Image,
+    Text
+} from 'react-native';
 var SimpleMarkdown = require('simple-markdown');
 var _ = require('lodash');
 
